### ISSUE TITLE
[RFC]: add W8A16 support

### ISSUE THREAD

**Issue by realliujiaxu** [vllm-project/vllm-ascend#4267]
[中文在线文档](https://docs.google.com/document/d/1yPE-YSqW080u6ifi3d0Or0Y89qKwddOoivipvbeusjc/edit?usp=sharing)
### Motivation.

Currently, vllm-ascend already supports various quantization precisions, such as W8A8, W4A8, and W4A4. Compared to these, the W8A16 (8-bit weights, 16-bit activations) configuration offers the following advantages:

- Higher precision. Quantization can typically be quickly completed using the AWQ algorithm, which ensures precision and requires less debugging effort.

- In low-concurrency scenarios, W8A16 offers a significant advantage over W16A16 (FP16): It loads less weight data, which enables faster inference speed in memory-bound scenarios. Compared to W8A8 (full quantization), W8A16 does not require the insertion of extra activation quantization/dequantization operators, thus avoiding additional computation overhead, resulting in faster inference speed.

### 1. Add AscendW8A16LinearMethod
The following is a pseudo-code implementation:

```Python

class AscendW8A16LinearMethod:
    def __init__(self) -> None:
        # According to the documentation (https://www.hiascend.com/document/detail/zh/Pytorch/720/apiref/torchnpuCustomsapi/context/torch_npu-npu_weight_quant_batchmatmul.md),
        # when the data format is ND, it is recommended to use the transposed weight as input in per-channel scenarios to improve performance.
        self.transpose_weight = ...

    @staticmethod
    def get_weight(
        input_size: int,
        output_size: int,
        params_dtype: torch.dtype = torch.bfloat16,
    ) -> Dict[str, Any]:
        params_dict = {
            "weight": torch.empty(output_size, input_size, dtype=torch.int8)
        }
        return params_dict

    @staticmethod
    def get_perchannel_param(
        output_size: int,
        params_dtype: torch.dtype,
    ) -> Dict[str, Any]:
        params_dict = {}
        params_dict["weight_scale"] = torch.empty(output_size,
                                                  1,
                                                  dtype=params_dtype)
        params_dict["weight_offset"] = torch.empty(output_size,
                                                   1,
                                                   dtype=params_dtype)
        return params_dict

    def get_pergroup_param(self,
                           input_size: int,
                           output_size: int,
                           params_dtype: torch.dtype,
                           layer_type: Optional[str] = None) -> Dict[str, Any]:
        # Mimic get_perchannel_param to return per-group weight_scale and weight_offset.
        return {}

    @staticmethod
    def apply(
        layer: torch.nn.Module,
        x: torch.Tensor,
        bias: Optional[torch.Tensor] = None,
        tp_rank: Optional[int] = 0,
    ) -> torch.Tensor:
        return torch_npu.npu_weight_quant_batchmatmul(
            x, layer.weight, layer.weight_scale, layer.weight_offset)

    def process_weights_after_loading(self, layer):
        if self.transpose_weight:
            layer.weight.data = layer.weight.data.transpose(0, 1).contiguous()
        if is_enable_nz():
            layer.weight.data = torch_npu.npu_format_cast(
                layer.weight.data, ACL_FORMAT_FRACTAL_NZ)
        layer.weight_scale.data = torch.flatten(layer.weight_scale.data)
        layer.weight_offset.data = torch.flatten(layer.weight_offset.data)
```

### 2. Register in ASCEND_QUANTIZATION_METHOD_MAP
It must be noted that the key needs to correspond to the model_quant_type field in the quantified weight's quant_model_description.json.

The registered value should be `{"linear": AscendW8A16LinearMethod}`.

**Comment by dongxi88** [comment]
合入到0.18.0了吗？

**Comment by lizy124** [comment]
> [中文在线文档](https://docs.google.com/document/d/1yPE-YSqW080u6ifi3d0Or0Y89qKwddOoivipvbeusjc/edit?usp=sharing)
> 
> ### Motivation.
> Currently, vllm-ascend already supports various quantization precisions, such as W8A8, W4A8, and W4A4. Compared to these, the W8A16 (8-bit weights, 16-bit activations) configuration offers the following advantages:
> 
> * Higher precision. Quantization can typically be quickly completed using the AWQ algorithm, which ensures precision and requires less debugging effort.
> * In low-concurrency scenarios, W8A16 offers a significant advantage over W16A16 (FP16): It loads less weight data, which enables faster inference speed in memory-bound scenarios. Compared to W8A8 (full quantization), W8A16 does not require the insertion of extra activation quantization/dequantization operators, thus avoiding additional computation overhead, resulting in faster inference speed.
> 
> ### Proposed Change.
> ### 1. Add AscendW8A16LinearMethod
> The following is a pseudo-code implementation:
> 
> class AscendW8A16LinearMethod:
>     def __init__(self) -> None:
>         # According to the documentation (https://www.hiascend.com/document/detail/zh/Pytorch/720/apiref/torchnpuCustomsapi/context/torch_npu-npu_weight_quant_batchmatmul.md),
>         # when the data format is ND, it is recommended to use the transposed weight as input in per-channel scenarios to improve performance.
>         self.transpose_weight = ...
> 
>     @staticmethod
>     def get_weight(
>         input_size: int,
>         output_size: int,
>         params_dtype: torch.dtype = torch.bfloat16,
>     ) -> Dict[str, Any]:
>         params_dict = {
>             "weight": torch.empty(output_size, input_size, dtype=torch.int8)
>         }
>         return params_dict
> 
>     @staticmethod
>     def get_perchannel_param(
>         output_size: int,
>         params_dtype: torch.dtype,
>     ) -> Dict[str, Any]:
>         params_dict = {}
>         params_dict["weight_scale"] = torch.empty(output_size,
>                                                   1,
>                                                   dtype=params_dtype)
>         params_dict["weight_offset"] = torch.empty(output_size,
>                                                    1,
>                                                    dtype=params_dtype)
>         return params_dict
> 
>     def get_pergroup_param(self,
>                            input_size: int,
>                            output_size: int,
>                            params_dtype: torch.dtype,
>                            layer_type: Optional[str] = None) -> Dict[str, Any]:
>         # Mimic get_perchannel_param to return per-group weight_scale and weight_offset.
>         return {}
> 
>     @staticmethod
>     def apply(
>         layer: torch.nn.Module,
>         x: torch.Tensor,
>         bias: Optional[torch.Tensor] = None,
>         tp_rank: Optional[int] = 0,
>     ) -> torch.Tensor:
>         return torch_npu.npu_weight_quant_batchmatmul(
>             x, layer.weight, layer.weight_scale, layer.weight_offset)
> 
>     def process_weights_after_loading(self, layer):
>         if self.transpose_weight:
>             layer.weight.data = layer.weight.data.transpose(0, 1).contiguous()
>         if is_enable_nz():
>             layer.weight.data = torch_npu.npu_format_cast(
>                 layer.weight.data, ACL_FORMAT_FRACTAL_NZ)
>         layer.weight_scale.data = torch.flatten(layer.weight_scale.data)
>         layer.weight_offset.data = torch.flatten(layer.weight_offset.data)
> ### 2. Register in ASCEND_QUANTIZATION_METHOD_MAP
> It must be noted that the key needs to correspond to the model_quant_type field in the quantified weight's quant_model_description.json.
> 
> The registered value should be `{"linear": AscendW8A16LinearMethod}`.
> 
> ### Feedback Period.
> _No response_
> 
> ### CC List.
> _No response_
> 
> ### Any Other Things.
> _No response_

Thank you for contributing the W8A16 quantization support!

The feature has been successfully merged via PR #4541 (commit `5018f2d8`) and is available since v0.13.0, with official release in v0.18.0.

**Implementation details:**
- ✅ Core implementation: `vllm_ascend/quantization/methods/w8a16.py`
- ✅ Registered in quantization methods
- ✅ Tests included: `tests/ut/quantization/methods/a2/test_w8a16.py`
- ✅ Documented in release notes

Closing this RFC as completed. Thanks again for the contribution!
